### PR TITLE
Add variables for substitutions on openid_connect_redirect_url

### PR DESCRIPTION
**Pull request by mayrstefan**
## Description

We have two Centreon-UIs using the same database and want to enable OpenID Connect for both instances. Originally both share the same redirect URL. We would like each instance to use the hostname the user connected to.

This patch adds three variables that can be used for substitution in the redirect URL:
* {$scheme} for http/https
* {$hostname} hostname or ip address the user used to access this Centreon-UI
* {$port} port used to access the Centreon-UI

So this can be formed to something like `{$scheme}://{$hostname}:{$port}/centreon/index.php`

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [ ] New functionality (non-breaking change)
- [x] Breaking change (patch or feature) that might cause side effects breaking part of the Software
- [ ] Updating documentation (missing information, typo...)

## Target serie

- [ ] 20.04.x
- [ ] 20.10.x
- [ ] 21.04.x
- [x] 21.10.x (master)

<h2> How this pull request can be tested ? </h2>

Use all variables as OpenID Connect Redirect URL or only some of them

## Checklist

- [ ] I have followed the **coding style guidelines** provided by Centreon
- [ ] I have commented my code, especially new **classes**, **functions** or any **legacy code** modified. (***docblock***)
- [ ] I have commented my code, especially **hard-to-understand areas** of the PR.
- [x] I have made corresponding changes to the **documentation** (help text is extended)
- [x] I have **rebased** my development branch on the base branch (master, maintenance).

### CODE REVIEW

**Comment by lpinsivy** [comment]
@mayrstefan should be better to use {$scheme}://{$hostname}:{$port}/centreon/index.php if the openid_connect_redirect_url field is empty, else use defined value.

Because what about a Centreon central server with [redundancy](https://docs.centreon.com/current/en/installation/installation-of-centreon-ha/overview.html#architectures) and VIP?

**Comment by mayrstefan** [comment]
I added "{$scheme}://{$hostname}:{$port}/centreon/index.php" as default value if this option is empty but left the possibility to substitute any of the mentioned variables if necessary.

About the redundancy question: we already have a redundant central server with failover VIPs. The web UI on this server is only used by central IT staff. All other centreon users access the separated UI instance that can be disconnected if the central server gets overloaded.

**Comment by lpinsivy** [comment]
HI @mayrstefan , thank you for your changes.

Just last one, can you change '/centreon' by $centreon->optGen["oreon_web_path"].
This is the field defined in "Administration > Parameters > Centreon Web Directory".

Regards,

File: www/class/centreonAuth.SSO.class.php [hunk 1]
```diff
@@ -103,7 +102,17 @@ public function __construct(
             # Get configured values
             $clientId = $this->ssoOptions['openid_connect_client_id'];
             $clientSecret = $this->ssoOptions['openid_connect_client_secret'];
-            $redirectNoEncode = $this->ssoOptions['openid_connect_redirect_url'];
+            if (empty($this->ssoOptions['openid_connect_redirect_url'])) {
+                $redirectNoEncode = '{$scheme}://{$hostname}:{$port}/centreon/index.php';
```
Comment (lpinsivy) [inline www/class/centreonAuth.SSO.class.php:106]: $redirectNoEncode = '{$scheme}://{$hostname}:{$port}' . rtrim($centreon->optGen["oreon_web_path"], "/") . '/index.php';

File: www/include/Administration/parameters/general/help.php [hunk 2]
```diff
@@ -178,7 +178,8 @@
 );
 $help['openid_connect_redirect_url'] = dgettext(
     'help',
-    'Your OpenId Connect redirect url (this server).'
+    'Your OpenId Connect redirect url (this server, {$scheme}, {$hostname} and {$port} can be used for substitions,'
+    . ' default is {$scheme}://{$hostname}:{$port}/centreon/index.php if left empty).'
```
Comment (lpinsivy) [inline www/include/Administration/parameters/general/help.php:182]: . ' default is {$scheme}://{$hostname}:{$port}/your_centreon_path/index.php if left empty).'

**Review by lpinsivy — CHANGES_REQUESTED** [review]

**Comment by lpinsivy** [comment]
@mayrstefan moreover, how did you configure your Service Provider? Did you create 2 separated configuration or did you allowed both IP address of Centreon servers?

**Comment by lpinsivy** [comment]
Internal Ticket MON-10862

File: www/class/centreonAuth.SSO.class.php (hunk 1 above)
Comment (mayrstefan) [reply www/class/centreonAuth.SSO.class.php]: Oreon - on the trail of history

I was already looking for something like that. Your hint is good but had another challenge: $centreon is not available during this state of authentication. The Centreon constructor always requires a userinfo to create an object. It took me a while until I realized I have to query the database for it. Like it is done in centreonLog.class.php. Maybe another thing that should be refactored one day. But not in this pull request.

**Comment by mayrstefan** [comment]
@lpinsivy for simple testing you can use the DNS name and the IP address of your Centreon instance and configure both of them as valid redirect URLs in Keycloak. Before this change you had to decide which one to use with OIDC. With this patch it is up to your SP configuration what you want to allow.

This mimics our setup: 
- two Centreon UIs with different DNS names
- one SP with one configured OpenID Connect ClientId for Centreon
- two whitelisted redirect URLs for this ClientId

File: www/class/centreonAuth.SSO.class.php [hunk 3]
```diff
@@ -103,7 +102,27 @@ public function __construct(
             # Get configured values
             $clientId = $this->ssoOptions['openid_connect_client_id'];
             $clientSecret = $this->ssoOptions['openid_connect_client_secret'];
-            $redirectNoEncode = $this->ssoOptions['openid_connect_redirect_url'];
+            if (empty($this->ssoOptions['openid_connect_redirect_url'])) {
+                /*
+                 * we can not access $centreon->optGen["oreon_web_path"] before authentication and have to query the
+                 * DB directly
+                 */
+                $pathSql = "SELECT `value` FROM `options` WHERE `key` = 'oreon_web_path'";
```
Comment (lpinsivy) [inline www/class/centreonAuth.SSO.class.php:110]: Hi @mayrstefan I think you can remove your SQL query and use $this->ssoOptions['oreon_web_path'];

The CentreonAuthSSO is created with $generalOptions var and this table contains all entries from centreon.options table.
And $this->ssoOptions is a copy of $generalOptions

**Review by lpinsivy — CHANGES_REQUESTED** [review]

File: www/class/centreonAuth.SSO.class.php [hunk 4]
```diff
@@ -103,7 +102,18 @@ public function __construct(
             # Get configured values
             $clientId = $this->ssoOptions['openid_connect_client_id'];
             $clientSecret = $this->ssoOptions['openid_connect_client_secret'];
-            $redirectNoEncode = $this->ssoOptions['openid_connect_redirect_url'];
+            if (empty($this->ssoOptions['openid_connect_redirect_url'])) {
+                $redirectNoEncode = '{$scheme}://{$hostname}:{$port}'
+                . $this->ssoOptions['oreon_web_path'] . 'index.php';
```
Comment (lpinsivy) [inline www/class/centreonAuth.SSO.class.php:107]: use rtrim to remove potential '/' defined by oreon_web_path and add a "/" in concatenation

Comment (lpinsivy) [reply www/class/centreonAuth.SSO.class.php:107]: fix with PR #10098 10098

**Comment by lpinsivy** [comment]
Replaced by PR #10098 10098

**Pull request closed without merging**